### PR TITLE
UTF8 Marshaling APIs 

### PR DESCRIPTION
**Pull request by migueldeicaza**
Adds support for the new UTF8 marshaling functionality in .NET, there is support for the new MarshalAs(..UTF8) as well as a handful of convenience methods in Marshal.cs

This fixes a few bugs that the new test suite exhibited: when we marshaled UTF8 strings, we assumed Chars() == Bytes() and we would end up chopping when marshaling out, and marshaling in. 

There is also an additional behavioral bug that was fixed in Marshal.cs, unlike the rest of the runtime that treated Ansi as Utf8, StringToCoTaskMemAnsi behaved as ascii.   This brings the API in line with the rest of the runtime.

The equivalent CoreCLR changes were:

* https://github.com/dotnet/coreclr/pull/4793
* https://github.com/dotnet/coreclr/pull/6561

This has also surfaced two limitations in Mono's marshaling code, which is why two sets of tests are disabled: StringBuilder return types are not handled specially (char * needs to be turned into a new StringBuilder when present) and StringBuilders with an Out attribute should update the StringBuilder in place, not make a new copy of it.

### CODE REVIEW

File: mono/tests/Makefile.am [hunk 1]
```diff
@@ -193,6 +193,7 @@ BASE_TEST_CS_SRC_UNIVERSAL=		\
 	typeof-ptr.cs		\
 	static-constructor.cs	\
 	pinvoke.cs		\
+	pinvoke-utf8.cs		\
```
Comment (kumpera) [inline mono/tests/Makefile.am]: This file is missing.

Comment (migueldeicaza) [inline mono/tests/Makefile.am]: Good catch!   Just added it.

**Comment by migueldeicaza** [comment]
build

**Comment by migueldeicaza** [comment]
build

**Comment by duncanmak** [comment]
build

**Comment by migueldeicaza** [comment]
build

**Pull request merged**